lib/Default/SmrPlanet.class.inc

```diff
@@ -982,20 +979,37 @@ class SmrPlanet {
 		return true;
 	}
 
+	// Modifier for planet building based on the number of buildings.
+	// The average value of this modifier should roughly be 1.
+	private function getCompletionModifier($constructionID) {
+		$currentBuildings = $this->getBuilding($constructionID);
+		$maxBuildings = $this->getMaxBuildings($constructionID);
+		return 0.01 + 2.97 * pow($currentBuildings/$maxBuildings, 2);
+	}
+
+	// Amount of exp gained to build the next building of this type
+	function getConstructionExp($constructionID) {
+		$baseExp = 0;
+		$this->db->query('SELECT exp_gain FROM planet_can_build WHERE planet_type_id=' . $this->db->escapeNumber($this->getTypeID()) . ' AND construction_id=' . $this->db->escapeNumber($constructionID));
+		if ($this->db->nextRecord()) {
+			$baseExp = $this->db->getInt('exp_gain');
+		}
+		$expGain = ceil($baseExp * $this->getCompletionModifier($constructionID));
+		return $expGain;
+	}
+
+	// Amount of time (in seconds) to build the next building of this type
 	function getConstructionTime($constructionID) {
 		$baseTime = 0;
 		$this->db->query('
```

Comment:
Should this also use the `Globals::getPlanetBuildings()` to share the cache in multiple places (if it's reused more places then the cost is much lower)